pkg/driver/controller.go

```diff
@@ -0,0 +1,359 @@
+			if err := d.mounter.MakeDir(target); err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not create dir %q: %v", target, err)
+			}
+			if err := d.mounter.Mount(fileSystemId, target, "efs", mountOptions); err != nil {
+				os.Remove(target)
+				return nil, status.Errorf(codes.Internal, "Could not mount %q at %q: %v", fileSystemId, target, err)
+			}
+			os.RemoveAll(target + accessPoint.AccessPointRootDir)
+			err = d.mounter.Unmount(target)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not unmount %q: %v", target, err)
+			}
+			os.RemoveAll(target)
+		}
+
+		// Delete access point
+		if err = d.cloud.DeleteAccessPoint(ctx, accessPointId); err != nil {
+			if err == cloud.ErrAccessDenied {
+				return nil, status.Errorf(codes.Unauthenticated, "Access Denied. Please ensure you have the right AWS permissions: %v", err)
+			}
+			if err == cloud.ErrNotFound {
+				klog.V(5).Infof("DeleteVolume: Access Point not found, returning success")
+				return &csi.DeleteVolumeResponse{}, nil
+			}
+			return nil, status.Errorf(codes.Internal, "Failed to Delete volume %v: %v", volId, err)
+		}
+	} else {
+		return nil, status.Errorf(codes.NotFound, "Failed to find access point for volume: %v", volId)
+	}
+
+	return &csi.DeleteVolumeResponse{}, nil
+}
+
+func (d *Driver) ControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (d *Driver) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+func (d *Driver) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
+	klog.V(4).Infof("ValidateVolumeCapabilities: called with args %+v", *req)
+	volId := req.GetVolumeId()
+	if volId == "" {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID not provided")
+	}
+
+	volCaps := req.GetVolumeCapabilities()
+	if len(volCaps) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities not provided")
+	}
+
+	fsId, _, accessPointId, err := parseVolumeId(volId)
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "Volume not found, err: %v", err)
+	}
+
+	if _, err := d.cloud.DescribeFileSystem(ctx, fsId); err != nil {
+		if err == cloud.ErrAccessDenied {
+			return nil, status.Errorf(codes.Unauthenticated, "Access Denied. Please ensure you have the right AWS permissions: %v", err)
+		}
+		if err == cloud.ErrNotFound {
+			return nil, status.Error(codes.NotFound, "Volume not found")
+		}
+		return nil, status.Errorf(codes.Internal, "Could not get volume with ID %q: %v", volId, err)
+	}
+
+	if accessPointId != "" {
+		if _, err := d.cloud.DescribeAccessPoint(ctx, accessPointId); err != nil {
```

Comment:
making cloud calls here is not necessary, this function should just be concerned with volume capabilities like SINGLE_NODE_WRITER, spec here https://github.com/container-storage-interface/spec/blob/master/spec.md#validatevolumecapabilities

Reply:
ah thanks for pointing this out, I'll update the function accordingly